### PR TITLE
Increase TaskID size by 2 bytes, taken from JobID

### PR DESCRIPTION
**Pull request by edoakes**
## Why are these changes needed?

Our TaskIDs are too short - there is a ~25% likelihood of a collision after 1e7 tasks submitted from the same 
task. This change shaves off two bytes from the JobID and allocates them to the task ID. Given that these are sequentially assigned, that means we can currently support 65k drivers.

Previously, we deterministically saw a collision after ~2e6 TaskIDs were generated from the driver (was causing failures in release testing), now there are none after 1e8.

We could probably shave off quite a few bits from the ObjectID flags as well, but that would require a larger change as they would no longer be byte-aligned.

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://ray.readthedocs.io/en/latest/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failure rates at https://ray-travis-tracker.herokuapp.com/.

### CODE REVIEW

File: src/ray/common/id.h [hunk 1]
```diff
@@ -99,7 +99,7 @@ class UniqueID : public BaseID<UniqueID> {
 
 class JobID : public BaseID<JobID> {
  public:
-  static constexpr int64_t kLength = 4;
+  static constexpr int64_t kLength = 2;
 
   static JobID FromInt(uint32_t value);
```
Comment (pcmoritz) [inline src/ray/common/id.h]: Let's make this `uint16_t` and add checks at where it is used to protect against silent overflow.

**Review by pcmoritz — APPROVED** [review]

**Review by ericl — APPROVED** [review]

**Review by ericl — APPROVED** [review]

**Review by raulchen — COMMENTED** [review]
Can you also update `id_specification.md`?

File: src/ray/common/id.h (hunk 1 above)
Comment (edoakes) [reply src/ray/common/id.h]: Done

File: src/ray/design_docs/id_specification.md [hunk 2]
```diff
@@ -26,14 +26,14 @@ Ray ID Specification
 +---------------------------+---------------------------------------------------------------+
 
 ```
-#### JobID (4 bytes)
-`JobID` is generated by `GCS` to ensure uniqueness. Its length is 4 bytes.
+#### JobID (2 bytes)
+`JobID` is generated by `GCS` to ensure uniqueness. Its length is 2 bytes.
 
 #### ActorID (8 bytes)
```
Comment (kfstorm) [inline src/ray/design_docs/id_specification.md]: `ActorID` should be 6 bytes, right?

**Pull request merged**